### DISCUSSION TITLE
Cross-cutting containers?

### DISCUSSION THREAD

**Discussion by azban** [terrastruct/d2#328] [Feature Requests]
Hi,
I'm looking to overlap containers, rather than requiring containers to be nested.

![image](https://user-images.githubusercontent.com/714931/205410586-06b4c8de-3107-43ae-9f41-8f22e054100e.png)

I was hoping that I could have the "cloud" container intersect the "app" container and have app.server live in the intersection. Is this possible? Is there an alternative that people like?

```d2
app: {
  client: {
    shape: image
    icon: https://icons.terrastruct.com/tech/browser-2.svg
  }

  server: {
    icon: https://icons.terrastruct.com/azure/Compute%20Service%20Color/VM/VM-windows.svg
    shape: image
  }

  client -> server
}

cloud: {
  app.server

  database
}
```

**Comment by alixander** [comment]
yeah this is not possible in D2 right now. i agree this should be possible. will think on the syntax...

**Comment by pleshevskiy** [comment]
Maybe I misunderstood you, but maybe you wanted something like this?

![image](https://user-images.githubusercontent.com/7839004/205913333-db56a983-ad60-4b71-9fa6-58438d6743bd.png)


```
app {
  client: {
    shape: image
    icon: https://icons.terrastruct.com/tech/browser-2.svg
  }
}

app.client -> cloud.app.server

cloud: {
  app: {
    server: {
      icon: https://icons.terrastruct.com/azure/Compute%20Service%20Color/VM/VM-windows.svg
      shape: image
    }
  }

  database
}
```

**Reply by azban** [reply]
Nope, looking for something like this:

![IMG_20221206_105133.jpg](https://user-images.githubusercontent.com/714931/205985451-e3829cfa-a922-49af-960a-41450bd60ab1.jpg)

**Reply by pleshevskiy** [reply]
Oh, very interesting. Haven't seen that before.

**Reply by brandonzylstra** [reply]
Isn't the DB a part of the app too? 

I realize this is simply an example, and there may be other examples where it makes sense for the "cloud" box to be partially inside and partially outside the "app" box, but in this case I don't see the value of it.  

Also the whole idea of the "cloud" in diagrams is that there are lots of details we're ignoring.  What is the detail here that is so important to capture?  

I hope I don't sound like I'm being difficult.  I'm really just trying to understand the use case, and I'd really like to see an example where it makes sense to do this.

**Comment by alixander** [comment]
What if we allow defining shapes that can specify encompassing other objects.

```d2
x.y -> a.b

c: {
  # c would be drawn to be around the bounding box of x.y and a
  constraint: [x.y; a]
}
```
I think we can reuse the `constraint` keyword, it's unambiguous and sounds reasonable. 

The limitation is that these shapes themselves cannot have any children.

**Reply by alixander** [reply]
E.g. for the above example here: https://github.com/terrastruct/d2/discussions/328#discussioncomment-4325624

```d2
app: {
  client
  server
}
db

cloud: {
  constraint: [db; app.server]
}
```

**Reply by alixander** [reply]
cc @gavin-ts

**Reply by gavin-ts** [reply]
That sounds reasonable, although it may not always be possible to only cover the specified shapes.

If we have these "surround" shapes that just draw a shape around (the bounding box of) a list of shapes, other shapes may end up inside the region as well. This isn't necessarily a downside since you can just enter the corners instead of needing to specify all contents.

**Reply by arjunrbery** [reply]
How about something that utilizes a different syntax to traverse up? For example, in html
```
<img src="./currentDirectoryImage.jpg"/>

<img src="../parentDirectoryImage.jpg"/>

<img src="../../grandparentDirectoryImage.jpg"/>

```

```

app: {
  client
  server
}
db

cloud: {
  .app.server
  .db
}

```
Drafted from phone, sorry about the formatting.

**Reply by brandonzylstra** [reply]
This is extremely interesting.  I'm not 100% sure about _exactly_ this syntax, but something very much along these lines seems like a good way to go.

It seems like the `.` is not so much implying any upward traversal, so much as breaking normal nesting rules.

**Reply by gavin-ts** [reply]
noting we already can use `_` to reference a parent container
https://d2lang.com/tour/containers#reference-parent

**Reply by gavin-ts** [reply]
I think in the context of this new shape, it would automatically consider these as `_.app.server` and `_.db` because you wouldn't define these as children (otherwise it would just be a normal shape). We only need this feature for going across containers, so I think it makes sense to have it count one level up by default, but `_` could still be used to reference further upwards

```d2
project1: {
 app: {
   client
   server
 }
 db
 
 cloud: {
   constraint: [db; app.server]
 }
}

project2: {
  app
  db

  sharedCloud: {
    constraint: [_.project1.app, _.project1.db, app, db]
  }
}
```

**Reply by gavin-ts** [reply]
we also might be able to just use the same container body, and allow these to also be any shape

```d2
app: {
  client
  server
}
db

cloud: {
  shape: cloud
  constraint: surround
  db
  app.server
}
```

**Reply by gavin-ts** [reply]
This is a bit of a tangent, but could we make the parent reference `../` instead of `_.`? I don't remember why `_.` was selected, was it to avoid conflicts with local files or something? I would think we could distinguish the two given parent ref is used as a key, and files are used as a value

**Reply by alixander** [reply]
You're proposing changing the syntax for going into a container from `.` to `/` for this one use case?

**Reply by gavin-ts** [reply]
oh right that's why, it would be `../../grandparent` going up, but `grandparent.parent.child` going down

so `_._.grandparent` is more consistent with `grandparent.parent.child` (to be clear)

If we did change it, I think it should keep `parent.child` for downwards and only be `../uncle` going upwards, but I see the inconsistency, I think  `../` is just seems more natural to me than `_.`

**Reply by arjunrbery** [reply]
The `../../grandparent` feels intuitive, but the `_._.grandparent` looks way better in my opinion.

**Reply by brandonzylstra** [reply]
`../../grandparent` follows shell path conventions, where `/` is the separator and `.` refers to a current directory and `..` refers to a parent, while on the other hand `grandparent.parent.child` follows the dot-notation of object-oriented programming languages, where `.` is a separator.  Mixing the two together would be **_very_** confusing and offer no benefit.  Using `_` as a parent makes perfect sense to me, even if I don't know of any precedent in some other syntactic system for doing so.

**Comment by brandonzylstra** [comment]
I don't find `constraint` to convey the idea of encompassing other objects without following normal nesting rules.  How about this instead:

```
x.y -> a.b

c: {
  encompass: [x.y; a]
}
```

**Comment by alixander** [comment]
TALA does actually have the foundations for this. For example, the code is set up for a "place X  in a way that a bounding box between X and Y and Z do not include any other shapes"

**Comment by silopolis** [comment]
Hi,
Kindly pointed here by @alixander on Discord, I'm joining that thread to share my use case for this: being able to code diagrams like the one below:

![image](https://github.com/terrastruct/d2/assets/414540/35d854f4-3b8b-4e31-b6f4-80aa5af7d7e5)

Specifically, being able to represent the Auto-Scaling Groups and the DB subnet group that are crossing the availability zones groups.

About the syntax, I believe I could live with only absolute `root.path.to.entity` refs, allthemore if aliases can be defined when the need to shorten names becomes pressing. But, of course, relative path are surely nice to have and talking about them, contrary to first thought, I prefer the `_` notation.

Something like:

```d2
AZ1: {
  WebNet: {
	EC2_Web
  }
  AppNet: {
	EC2_App
  }
}

AZ2: {
  WebNet: {
	EC2_Web
  AppNet: {
	EC2_App
  }
}

ASG_Web: {
  AZ1.WebNet.EC2
  AZ2.WebNet.EC2
}

ASG_App: {
  AZ1.AppNet.EC2
  AZ2.AppNet.EC2
}

VPC: {
  AZ1.WebNet
  AZ2.WebNet
  AZ1.AppbNet
  AZ2.AppNet
}

Cloud: {
  AZ1
  AZ2
  VPC
}

```

On a nth look, the VPC case may be trickier :thinking:

Just hope this will eventually be part of the free tier as the subscription is not in my pocket reach...

Thanks for all
TY

**Comment by behniafb** [comment]
Is there any update on this feature?

**Comment by clipstick** [comment]
Can this be looked at as a feature, I would use this a lot if Cloud architecture diagrams and so would a lot of my colleagues.

**Comment by alixander** [comment]
Maybe we introduce a new `groups` keyword. 

```d2
groups: {
  aws_things: AWS Objects {
    style.fill: green
  }
}

x: {
  a.group: groups.aws_things
}

y: {
  b.group: groups.aws_things
}
```

-------------

Even though there's no way for Dagre or ELK to support this in a way that grouped things are close to each other for a clean box, in theory you can wrap the two nodes in some minimum bounding box or hoop.

Like what if you wanted to group satellites and storage here?

![Screenshot 2024-06-09 at 8 22 44 PM](https://github.com/terrastruct/d2/assets/3120367/8da1d6e8-2dfd-459b-9f18-82eb71340980)

This looks very unclean, but can't think of any other solutions and is maybe good enough. 

![Screenshot 2024-06-09 at 8 22 44 PM](https://github.com/terrastruct/d2/assets/3120367/af73943b-7b42-490e-9df6-f0ced57100da)

I guess you'd have to search a path from one corner of top-left-most object to the bottom-right-most that goes through all objects, and then inflate it. Seems very dubious if that can yield good results.

--------------

Another option is that this is a grid-specific feature.

Since grids is our layout algorithm, it can ensure that nodes end up close to each other. And from what I've seen of these use cases, it looks quite grid-like. 

This of course runs a big limitation that it won't work on regular objects in the main layout. Plus if you want to do this between two grids which are in separate containers which go through the main layout algorithm, it still can't do anything about proximity.

**Reply by databrecht** [reply]
I think it would be tremendously useful in other diagrams too, so I do hope that D2 considers a solution for all diagrams since the real world is simply not hierarchical. For example, currently I'm looking at a way to express sequence diagrams where you have flows that overlap. One request (r1) is not yet finished and another one (r2) starts as **part** of that request. It's not hierarchy, it's more of a spin-off. In bars that would be expressed as follows (which is also not possible if I'm not mistaken:
![CleanShot 2025-04-03 at 11 00 50](https://github.com/user-attachments/assets/84b4d3a3-1ca9-45f1-b285-a429f0069c20)
We could also try to express is via containers (which are used for Scoping https://d2lang.com/tour/sequence-diagrams/) but there we have a similar problem as above, we'd require partially overlapping containers. Below is a real-life example (with all api names removed)
![CleanShot 2025-04-03 at 11 09 17](https://github.com/user-attachments/assets/d81b2d64-dc85-46f2-b245-3126e4da7296)


I get that it's an algorithmic limitation due to the layout algorithms though, if Tala does provide the capabilities and it would be supported in D2, I would definitely buy Tala. It would overcome limitations of many other tools to generate diagrams.

**Comment by terlar** [comment]
Another example is visualization of multi-cluster setups:
![image](https://github.com/terrastruct/d2/assets/280235/c711a906-85f2-4f4e-bd1f-038de0a21fa3)

![image](https://github.com/terrastruct/d2/assets/280235/14004ad7-d54e-441b-b8be-d6ddd9ae4218)

**Comment by AndydeCleyre** [comment]
If for nothing else than to make this discussion turn up for some relevant search terms, I'll mention Venn diagrams and [Euler diagrams](https://en.wikipedia.org/wiki/Euler_diagram).

![image](https://github.com/user-attachments/assets/72012c13-8623-4859-ab8a-a51b55b237d2)

**Comment by GeorgeFischhof** [comment]
Hi,
I wanted this too: had a quite complex diagram (network like some people mentioned similar before), mermaid was not enough for that, and found D2. 

Found "classes" which is reused, and sub-graphs are accessed via hierarchy with dot notation, so I started to find the way to create "prototypes" of sub-graphs and use them in several containers. (but not found...)

I think technically it could be solved with composition:

```
apple: Apple {
  grid-columns: 1

  stem: {
    shape: diamond
  }

  friut: Fruit {
    shape: circle
    style.fill: red
  }

  stem -- friut
}

apple_tree: Apple Tree {
  tree: Tree {
    branch: Branch
    shape: cylinder
    branch -- apple.multiplicated
    branch -- apple.multiplicated
  }
}

```

so I do not use part of the apple, but create one apple, and use it several other diagrams, or for example with the `.multiplicated` keyword I use it in several times in one diagram. 

If used in several diagrams, D2 could duplicate it and rename to some unique id, and pass those multiplicated, renamed elements to the render engine

Note: for me it would not be a problem to compose the apple from fruit and stem if I would need to use the fruit or stem elsewhere.

BR,
George

**Reply by GeorgeFischhof** [reply]
Also it came to my mind that, it would be good to generate the graph from multiple files:
one (or more) could contain the data, the other(s) contain the diagram:

for example I had several IP address in my diagram. 

One file could contain all the 
- network names
- the machines in them
- and the IP address of the machines

for example:

```
# network_elements.d2
test_network {
    machine_1 {
        label: 127.0.0.1
    }
    machine_2 ...
...
}
```
```
# network_diagram.d2
# need some import mechanism...
import network_elements

full_network {
    test_network -> gateway_for_example
...
}
```

**Comment by Timouri** [comment]
Is there any update on this feature?

**Reply by brettcacu** [reply]
[Source](https://d2lang.com/tour/faq/#can-an-object-be-part-of-more-than-1-container)

> Not currently and not in the near future.

Found this snipped on the main FAQ.
